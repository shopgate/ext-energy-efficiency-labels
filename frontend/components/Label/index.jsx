import React from 'react';
import PropTypes from 'prop-types';
import Link from '@shopgate/pwa-common/components/Router/components/Link';
import styles from './style';

/**
 * Energy efficiency label
 * @param {Object} props Props
 * @return {JSX}
 */
const Label = ({ energyInfo, show }) => {
  if (!show || !energyInfo.class) {
    return null;
  }

  const index = energyInfo.class.indexOf('+');
  const text = index >= 0 ? energyInfo.class.substr(0, index) : energyInfo.class;
  const plus = index >= 0 ? energyInfo.class.substr(index) : '';

  const content = [text, <sub className={styles.sup}>{plus}</sub>];

  if (energyInfo.link) {
    return (
      <Link href={energyInfo.link} className={`${styles.arrow} ${styles.getColor(energyInfo.class)}`}>
        {content}
      </Link>
    );
  }

  return (
    <div className={`${styles.arrow} ${styles.getColor(energyInfo.class)}`}>
      {content}
    </div>
  );
};

Label.propTypes = {
  energyInfo: PropTypes.shape({
    class: PropTypes.string,
    link: PropTypes.string,
  }),
  show: PropTypes.bool,
};

Label.defaultProps = {
  energyInfo: {},
  show: true,
};

export default Label;
