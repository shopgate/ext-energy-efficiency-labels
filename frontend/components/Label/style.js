import { css } from 'glamor';
import getColorForClass from '../../helpers/colorMapping';

const arrow = css({
  position: 'relative',
  width: 34,
  height: 18,
  lineHeight: '18px',
  padding: '0 1px',
  margin: '6px 0 0 auto',
  fontSize: 16,
  fontWeight: 600,
  color: '#FFF',
  textAlign: 'center',
  ':before': {
    content: ' ',
    position: 'absolute',
    display: 'block',
    borderTop: '6.5px solid transparent',
    borderBottom: '6.5px solid transparent',
    borderRight: 'solid currentColor',
    borderWidth: '9px',
    borderLeft: 'none',
    left: 'auto',
    right: '100%',
  },
});

const sup = css({
  top: '-.25em',
  fontSize: '75%',
  lineHeight: 0,
  position: 'relative',
  verticalAlign: 'baseline',
});

/**
 * Returns the color styling for the given energy class
 * @param {string} energyClass Energy class
 * @return {string}
 */
const getColor = (energyClass) => {
  const color = getColorForClass(energyClass);

  return css({
    backgroundColor: color,
    ':before': {
      borderRightColor: color,
    },
  }).toString();
};

export default {
  arrow,
  getColor,
  sup,
};
