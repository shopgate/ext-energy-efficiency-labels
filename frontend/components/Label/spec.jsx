import React from 'react';
import { mount } from 'enzyme';
import Label from './';

// eslint-disable-next-line require-jsdoc, react/prop-types
const MockedLink = props => <div>{props.children}</div>;
jest.mock('@shopgate/pwa-common/components/Link', () => props => <MockedLink {...props} />);

describe('<Label />', () => {
  describe('No data available', () => {
    it('should render nothing', () => {
      const wrapper = mount(<Label />);

      expect(wrapper.html()).toBe(null);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Data available', () => {
    const energyInfo = {
      class: 'B',
      link: 'https://foo.bar',
    };

    it('should render nothing if show=false', () => {
      const wrapper = mount(<Label show={false} energyInfo={energyInfo} />);

      expect(wrapper.html()).toBe(null);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render energyInfo with <Link>', () => {
      const wrapper = mount(<Label energyInfo={energyInfo} />);

      expect(wrapper.find('MockedLink').exists()).toBe(true);
      expect(wrapper.find('MockedLink').prop('href')).toBe(energyInfo.link);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render energyInfo without <Link>', () => {
      const wrapper = mount(<Label energyInfo={{ class: 'A' }} />);

      expect(wrapper.find('MockedLink').exists()).toBe(false);
      expect(wrapper).toMatchSnapshot();
    });

    it('should split the "++"', () => {
      const wrapper = mount(<Label energyInfo={{ class: 'A++' }} />);

      expect(wrapper.text()).toBe('A++');
      expect(wrapper.find('sub').text()).toBe('++');

      expect(wrapper).toMatchSnapshot();
    });
  });
});
