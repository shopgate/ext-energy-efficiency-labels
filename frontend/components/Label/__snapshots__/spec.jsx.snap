// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Label /> Data available should render energyInfo with <Link> 1`] = `
<Link
  className="css-s5op6a css-1xzbir7"
  disabled={false}
  href="https://foo.bar"
>
  B
  <sub
    className={
      Object {
        "data-css-1xwhrfe": "",
      }
    }
  />
</Link>
`;

exports[`<Label /> Data available should render energyInfo without <Link> 1`] = `
<div
  className="css-s5op6a css-1xzbir7"
>
  A
  <sub
    className={
      Object {
        "data-css-1xwhrfe": "",
      }
    }
  />
</div>
`;

exports[`<Label /> Data available should render nothing if show=false 1`] = `""`;

exports[`<Label /> Data available should split the "++" 1`] = `
<div
  className="css-s5op6a css-11dk3ro"
>
  A
  <sub
    className={
      Object {
        "data-css-1xwhrfe": "",
      }
    }
  >
    ++
  </sub>
</div>
`;

exports[`<Label /> No data available should render nothing 1`] = `""`;
